internal/parser/parser.go

```diff
@@ -1049,6 +1055,79 @@ func (p *parser) repeat(fn func() bool) {
 	}
 }
 
+// open will open a new block. It will expect that the next token
+// is the start token and mark that we expect the end token in the
+// future.
+func (p *parser) open(start, end token.Token) (pos token.Pos, lit string) {
+	pos, lit = p.expect(start)
+	p.stack[end]++
+	return pos, lit
+}
+
+// more will check if we should continue reading tokens for the
+// current block. This is true when the next token is not EOF and
+// the next token is also not one that would close a block.
+func (p *parser) more() bool {
+	_, tok, _ := p.peek()
+	if tok == token.EOF {
+		return false
+	}
+	_, ok := p.stack[tok]
+	return !ok
```

Comment:
nit, but it would be clearer to me of this used the count and the zero value for if the token is not in the map (i.e. Go's default behavior)

```
count := p.stack[tok]
return count > 0
```